packages/babel-plugin-proposal-discard-binding/src/utils.ts

```diff
@@ -0,0 +1,146 @@
+    const parentHasInnerComments = innerComments.length > 0;
+    innerComments.push(...(leadingComments ?? []), ...(trailingComments ?? []));
+    if (parentHasInnerComments) {
+      innerComments.sort((a, b) => a.end - b.end);
+    } else {
+      parent.innerComments = innerComments;
+    }
+  }
+}
+
+function syncScopeForReplacedVoidPattern(path: NodePath<t.Identifier>) {
+  for (let childPath: NodePath = path; childPath.parentPath != null; ) {
+    const parentPath: NodePath = childPath.parentPath;
+    switch (parentPath.type) {
+      case "VariableDeclaration": {
+        const kind = parentPath.node.kind;
+        const bindingKind =
+          kind === "using" || kind === "await using" ? "const" : kind;
+        parentPath.scope.registerBinding(bindingKind, path, childPath);
+        return;
+      }
+      case "AssignmentExpression":
+      case "ForInStatement":
+      case "ForOfStatement":
+        parentPath.scope.push({ id: cloneNode(path.node) });
+        return;
+      case "CatchClause":
+        parentPath.scope.registerBinding("let", parentPath);
+        return;
+      default:
+        if (parentPath.isFunction()) {
+          parentPath.scope.registerBinding("param", path, childPath);
+          return;
+        }
+    }
+    childPath = parentPath;
+  }
+
+  throw new Error("@babel/plugin-proposal-discard-binding: Internal Error");
+}
+
+export function transformVoidPattern(
+  path: NodePath<t.VoidPattern>,
+): NodePath<t.Identifier> | null {
+  if (path.parentPath.isArrayPattern()) {
+    return transformVoidPatternIntoNull(path);
+  } else {
+    return transformVoidPatternIntoUidIdentifier(path);
+  }
+}
+
+function* iterateVoidPatternsInLVal(
+  path: NodePath<t.LVal>,
+): Generator<NodePath<t.VoidPattern>> {
+  switch (path.type) {
+    case "ArrayPattern":
+      for (const elementPath of path.get("elements")) {
+        yield* iterateVoidPatternsInLVal(elementPath);
+      }
+      break;
+    case "ObjectPattern":
+      for (const propertyPath of path.get("properties")) {
+        if (propertyPath.isObjectProperty()) {
+          yield* iterateVoidPatternsInLVal(
+            propertyPath.get("value") as NodePath<t.Pattern>,
+          );
+        }
+      }
+      break;
+    case "AssignmentPattern":
+      yield* iterateVoidPatternsInLVal(path.get("left"));
+      break;
+    case "VoidPattern":
+      yield path;
+      break;
+    default:
+      break;
+  }
+}
+
+function* iterateVoidPatternInVariableDeclaration(
+  path: NodePath<t.VariableDeclaration>,
+): Generator<NodePath<t.VoidPattern>> {
+  for (const declaratorPath of path.get("declarations")) {
+    yield* iterateVoidPatternsInLVal(declaratorPath.get("id"));
+  }
+}
+
+export function hasVoidPatternInVariableDeclaration(
+  path: NodePath<t.VariableDeclaration>,
+): boolean {
+  return !iterateVoidPatternInVariableDeclaration(path).next().done;
+}
+
+export function splitNamedDeclarationAsVarAndExport(
```

Comment:
This helper is also used in the `destructuring` and `object-rest-spread` transform. Do you think we should extract them into one `helper-destructuring-transform` package and reuse them in transforms? Or a mega `helper-create-destructuring-transform` package that lowers to different compilation targets?

Reply:
Maybe let's start wither with a small helper package, or by exporting it from `destructuring` and then making this plugin depend on it.